Source/BugsnagMetadata.m

```diff
@@ -43,81 +44,52 @@ - (id)init {
 
 - (id)initWithDictionary:(NSMutableDictionary *)dict {
     if (self = [super init]) {
-        self.dictionary = dict;
+        // Ensure that the instantiating dictionary is mutable.
+        // Saves checks later.
+        self.dictionary = [dict mutableDeepCopy];
     }
     [self.delegate metadataChanged:self];
     return self;
 }
 
-// MARK: - <NSMutableCopying>
-
-- (id)mutableCopyWithZone:(NSZone *)zone {
-    @synchronized(self) {
-        NSMutableDictionary *dict = [self.dictionary mutableCopy];
-        return [[BugsnagMetadata alloc] initWithDictionary:dict];
-    }
-}
-
-- (NSMutableDictionary *)getMetadata:(NSString *)sectionName {
-    @synchronized(self) {
-        return self.dictionary[sectionName];
-    }
-}
-
-- (NSMutableDictionary *)getMetadata:(NSString *)sectionName
-                                 key:(NSString *)key
+- (NSDictionary *)toDictionary
 {
     @synchronized(self) {
-        return [self.dictionary valueForKeyPath:[NSString stringWithFormat:@"%@.%@", sectionName, key]];
-    }
-}
-
-- (void)clearMetadataInSection:(NSString *)sectionName {
-    @synchronized(self) {
-        [self.dictionary removeObjectForKey:sectionName];
+        return [NSDictionary dictionaryWithDictionary:self.dictionary];
     }
-    [self.delegate metadataChanged:self];
 }
 
-- (void)clearMetadataInSection:(NSString *)section
-                           key:(NSString *)key
-{
+- (id)deepCopy {
     @synchronized(self) {
-        if ([[[self dictionary] objectForKey:section] objectForKey:key]) {
-            [[[self dictionary] objectForKey:section] removeObjectForKey:key];
-        }
+        return [[BugsnagMetadata alloc] initWithDictionary:[[self dictionary] mutableDeepCopy]];
     }
-    [self.delegate metadataChanged:self];
 }
 
-- (NSDictionary *)toDictionary {
-    @synchronized(self) {
-        return [NSDictionary dictionaryWithDictionary:self.dictionary];
-    }
-}
+// MARK: - <BugsnagMetadataStore>
 
 /**
  * Add a single key/value to a metadata Tab/Section.
  */
-- (void)addAttribute:(NSString *)attributeName
-           withValue:(id)value
-       toTabWithName:(NSString *)tabName {
-    
+- (void)addMetadata:(id)metadata
+            withKey:(NSString *)key
+          toSection:(NSString *)sectionName
+{
     bool metadataChanged = false;
     @synchronized(self) {
-        if (value && value != [NSNull null]) {
-            id cleanedValue = BSGSanitizeObject(value);
+        if (metadata && metadata != [NSNull null]) {
+            id cleanedValue = BSGSanitizeObject(metadata);
             if (cleanedValue) {
                 // Value is OK, try and set it
-                NSMutableDictionary *section = [self getMetadata:tabName];
+                NSMutableDictionary *section = [[self getMetadataFromSection:sectionName] mutableCopy];
```

Comment:
I believe the `mutableCopy` can be avoided here as we can guarantee that the section value is always an `NSMutableDictionary`, since we set it ourselves. The same applies to the 2-param `addMetadata` method.

Reply:
In which case the protocol signature also needs updated to include Mutable.

Reply:
Sounds like a good approach 👍